xbmc/pvr/PVRGUIActions.cpp

```diff
@@ -1843,22 +1843,29 @@ namespace PVR
     }
   }
 
+  void CPVRChannelSwitchingInputHandler::AppendChannelNumberCharacter(char cCharacter)
+  {
+    // special case. if only a single zero was typed in, switch to previously played channel.
+    if (GetCurrentDigitCount() == 0 && cCharacter == '0')
+    {
+      SwitchToPreviousChannel();
+      return;
+    }
+
+    CPVRChannelNumberInputHandler::AppendChannelNumberCharacter(cCharacter);
+  }
+
   void CPVRChannelSwitchingInputHandler::OnInputDone()
   {
     CPVRChannelNumber channelNumber;
-    bool bSwitchToPreviousChannel;
 
     {
       CSingleLock lock(m_mutex);
```

Comment:
@ksooo not sure if we still need the lock.

Reply:
We need it. `OnTimeout()`, which calls the input handler's worker `OnInputDone` is always called from timer thread.

<pre>
(lldb) thread backtrace
* thread #44, stop reason = breakpoint 1.1
  * frame #0: 0x0000000101abd390 kodi.bin`PVR::CPVRChannelNumberInputHandler::OnTimeout(this=0x0000000109127400) at PVRChannelNumberInputHandler.cpp:46
    frame #1: 0x0000000101fb0ab1 kodi.bin`void std::__1::__invoke_void_return_wrapper<void>::__call<std::__1::__bind<void (ITimerCallback::*)(), ITimerCallback*&>&>(std::__1::__bind<void (ITimerCallback::*)(), ITimerCallback*&>&&&) [inlined] decltype(__f=0x0000000109127ea8, __a0=0x0000000109127eb8)).*fp(std::__1::forward<>(fp1))) std::__1::__invoke<void (ITimerCallback::*&)(), ITimerCallback*&, void>(void (ITimerCallback::*&&&)(), ITimerCallback*&&&) at type_traits:4232
    frame #2: 0x0000000101fb0a32 kodi.bin`void std::__1::__invoke_void_return_wrapper<void>::__call<std::__1::__bind<void (ITimerCallback::*)(), ITimerCallback*&>&>(std::__1::__bind<void (ITimerCallback::*)(), ITimerCallback*&>&&&) [inlined] std::__1::__bind_return<void (ITimerCallback::*)(), std::__1::tuple<ITimerCallback*>, std::__1::tuple<>, __is_valid_bind_return<void (ITimerCallback::*)(), std::__1::tuple<ITimerCallback*>, std::__1::tuple<> >::value>::type std::__1::__apply_functor<void (__f=0x0000000109127ea8, __bound_args=0x0000000109127eb8, __args=0x0000700004ee3890)(), std::__1::tuple<ITimerCallback*>, 0ul, std::__1::tuple<> >(void (ITimerCallback::*&)(), std::__1::tuple<ITimerCallback*>&, std::__1::__tuple_indices<0ul>, std::__1::tuple<>&&) at functional:2206
    frame #3: 0x0000000101fb0a02 kodi.bin`void std::__1::__invoke_void_return_wrapper<void>::__call<std::__1::__bind<void (ITimerCallback::*)(), ITimerCallback*&>&>(std::__1::__bind<void (ITimerCallback::*)(), ITimerCallback*&>&&&) [inlined] std::__1::__bind_return<void (ITimerCallback::*)(), std::__1::tuple<ITimerCallback*>, std::__1::tuple<>, __is_valid_bind_return<void (ITimerCallback::*)(), std::__1::tuple<ITimerCallback*>, std::__1::tuple<> >::value>::type std::__1::__bind<void (this=0x0000000109127ea8)(), ITimerCallback*&>::operator()<>() at functional:2239
    frame #4: 0x0000000101fb09df kodi.bin`void std::__1::__invoke_void_return_wrapper<void>::__call<std::__1::__bind<void (ITimerCallback::*)(), ITimerCallback*&>&>(std::__1::__bind<void (ITimerCallback::*)(), ITimerCallback*&>&&&) [inlined] decltype(__f=0x0000000109127ea8)(), ITimerCallback*&>&>(fp)(std::__1::forward<>(fp0))) std::__1::__invoke<std::__1::__bind<void (ITimerCallback::*)(), ITimerCallback*&>&>(std::__1::__bind<void (ITimerCallback::*)(), ITimerCallback*&>&&&) at type_traits:4291
    frame #5: 0x0000000101fb09cb kodi.bin`void std::__1::__invoke_void_return_wrapper<void>::__call<std::__1::__bind<void (__args=0x0000000109127ea8)(), ITimerCallback*&>&>(std::__1::__bind<void (ITimerCallback::*)(), ITimerCallback*&>&&&) at __functional_base:359
    frame #6: 0x0000000101fb08e9 kodi.bin`std::__1::__function::__func<std::__1::__bind<void (ITimerCallback::*)(), ITimerCallback*&>, std::__1::allocator<std::__1::__bind<void (ITimerCallback::*)(), ITimerCallback*&> >, void ()>::operator(this=0x0000000109127ea0)() at functional:1552
    frame #7: 0x000000010053dceb kodi.bin`std::__1::function<void ()>::operator(this=0x0000000109127ea0)() const at functional:1903
    frame #8: 0x0000000101faf60b kodi.bin`CTimer::Process(this=0x0000000109127b60) at Timer.cpp:110
    frame #9: 0x0000000101fae212 kodi.bin`CThread::Action(this=0x0000000109127b60) at Thread.cpp:221
    frame #10: 0x0000000101fac831 kodi.bin`CThread::staticThread(data=0x0000000109127b60) at Thread.cpp:131
    frame #11: 0x00000001084d36b9 libsystem_pthread.dylib`_pthread_body + 340
    frame #12: 0x00000001084d3565 libsystem_pthread.dylib`_pthread_start + 377
    frame #13: 0x00000001084d2c55 libsystem_pthread.dylib`thread_start + 13
</pre>

 The other methods. like  `AppendChannelNumberCharacter`are called from gui thread.

Reply:
Wait! You are right. We don't need the mutex here as we are no longer calling different member functions in this block. `GetChannelNumber` protects itself. Good catch.

Reply:
Okay thanks .. removed.